apps/lib/apps.c

```diff
@@ -2377,42 +2374,61 @@ int raw_write_stdout(const void *buf, int siz)
  * and is therefore a show of intent more than anything else.  However, it
  * does impact behavior on some platforms, such as differentiating between
  * text and binary input/output on non-Unix platforms
+ *
+ * The routines below call check_bio() in case of failure, and therefore
+ * don't free up other resources on the error path because they know it will
+ * be handled when the process exits.
  */
+static BIO *check_bio(BIO *bp, const char *what, int mode)
```

Comment:
I've noticed that you're sending in `format` for mode in certain places.  They are not at all the same thing!
Also, I much prefer it if the error said text what was wanted with this particular BIO, i.e. pass a describing string, like the ones that `modeverb()` generated: "appending", "reading", "writing".  The error message could print like this:

``` C
        app_bail_out("%s: Could not open %s for %s\n",
                     opt_getprog(), what, reason);
```

Reply:
Yes, it's mixing number-spaces in a common error.  See the subthread at https://github.com/openssl/openssl/pull/11993#discussion_r434717452

Reply:
Yeah, and you misunderstand.  The original error message would say something like this:

> Can't open foo.something for writing, ...

("writing" could as well be "reading" or "appending")

With this PR it's saying:

> prog: Could not open foo.something in mode <magicnumber>

Which of the two do you think is more useful to the user?  Words like "writing", "appending", "reading", or a magic number?

Reply:
Sure, I get it.  (I am the one who wrote and used fmt2str, for example).  The problem with now is that mode and format are overloaded, as discussed upthread.  Untangle those, then fix the error messages.

Reply:
I have no idea why you would mix them in the first place.  If you want to be able to have both in this function (not sure I understand why), *then have them as two separate arguments*

Reply:
At first it was just a ```FMT_``` style argument.  Then during various reviews, others pointed out places where ```check_bio``` could be added, and it wasn't clear that this was an open(2) ```mode``` argument until other reviews pointed it out.

I'll get around to it shortly.

Reply:
You know what, I think it'll be easier to simply show you how I mean it.

Reply:
You don't have to spend any effort on this unless you don't have anything better to do.  I know what you and David want.  I didn't want to do it as part of this PR, but I will when I get time.  Or you can just push commits to this PR if you prefer.